memstore/list/memory_pool.go

```diff
@@ -0,0 +1,171 @@
+
+// HighLevelMemoryPool manages memory requests on pooled memory. It underlying uses a
+// slab allocator to manage free memory chunks. When it no longer can satisfy memory
+// allocation request from customer, it will request more memory from underlying
+// NativeMemoryPool. All address returned back to client are an 2 element array of offset
+// where the first offset is the memory allocated to the caller and second offset is the
+// footer offset to current slab.
+type HighLevelMemoryPool interface {
+	// Allocate allocates size byte memory and return back to client.
+	Allocate(size int) [2]uintptr
+	// Reallocate reallocates memory according to the size of the old allocated
+	// memory and size of new allocation requests. If the size is the same, it
+	// does nothing and just return the old addr. Otherwise it will allocate a new
+	// memory, copy the old content to it if oldSize is non-zero and returned back to
+	// client. If the oldSize is not zero, it will also free the old memory.
+	Reallocate(oldBuf [2]uintptr, oldSize int, newSize int) [2]uintptr
+	// Return the memory back to memory pool.
+	Free(buf [2]uintptr)
+	// Return the actual memory address given offset.
+	Interpret(offset uintptr) uintptr
+	// Release the underlying memory.
+	Destroy()
+}
+
+// slabMemoryPool implements the HighLevelMemoryPool using slab algorithm.
+type slabMemoryPool struct {
+	slabAllocator    *Arena
+	nativeMemoryPool NativeMemoryPool
+}
+
+// NewHighLevelMemoryPool returns a default implementation of HighLevelMemoryPool.
+func NewHighLevelMemoryPool(reporter HostMemoryReporter) HighLevelMemoryPool {
+	nativeMP := NewNativeMemoryPool(reporter)
+	slabAllocator :=
+		NewArena(defaultStartSlabAllocatorChunkSize, defaultSlabSize, defaultSlabGrowthFactor, nativeMP)
+	return slabMemoryPool{
+		slabAllocator:    slabAllocator,
+		nativeMemoryPool: nativeMP,
+	}
+}
+
+// Allocate is the implementation of Allocate in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Allocate(size int) [2]uintptr {
+	if size == 0 {
+		return [2]uintptr{0, 0}
+	}
+	return mp.slabAllocator.Alloc(size)
+}
+
+// Reallocate is the implementation of Reallocate in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Reallocate(oldBuf [2]uintptr, oldSize int, newSize int) [2]uintptr {
+	if oldSize == newSize {
+		return oldBuf
+	}
+
+	if oldSize != 0 {
+		mp.Free(oldBuf)
+	}
+
+	return mp.Allocate(newSize)
+}
+
+// Free is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Free(buf [2]uintptr) {
+	if !mp.slabAllocator.DecRef(buf) {
+		utils.GetLogger().Panic("buf should be freed but not")
+	}
+}
+
+// Interpret is the implementation of Interpret in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Interpret(offset uintptr) uintptr {
+	return offset + mp.nativeMemoryPool.GetBaseAddr()
+}
+
+// Destroy is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Destroy() {
```

Comment:
anything need to be done for `slabAllocator`?

Reply:
no, slabAllocator does not allocate any malloc any memory itself